### ISSUE TITLE
parmed vs xml: differences in energy

### ISSUE THREAD

**Issue by SHervoe-Hansen** [openmm/openmm#3643]
Hi again devs!

So in attempting to convert a force field in gromacs format to XML, I passed some behavior I like to run by you.

All the problems can be replicated from the files in [this tarball](https://github.com/openmm/openmm/files/8916621/topVSxml.tar.gz).

In specific I find 3 odd behaviors, which are most likely intertwined:
1. 
A single-point energy calculation on the same structure by the two formats yielded a significant change in potential energy. While there are rounding errors in the XML forcefield file, due to gromacs using degrees and openMM using radian, I would not imagine such a big energy difference.
```
Potential energy using peg.xml: 76.75855413311949 kJ/mol
Potential energy using PEG2.top: 81169.57595582199 kJ/mol
```
I fear I am staring myself blind to the problem. Can you detect anything out of the ordinary in the XML file?

2. I do not understand this output.
```python
print(system.getForces()[3].getExceptionParameters(0))
[4,
 0,
 Quantity(value=0.0, unit=elementary charge**2),
 Quantity(value=0.5, unit=nanometer),
 Quantity(value=0.0, unit=kilojoule/mole)]
```
This interaction is between C1 and H12, both of which possess non-zero LJ and charge parameters in the forcefield and with the fudge factors !=1 

3. Attempting to load the parmed topology into openMM and back into parmed yields differences in the fudge factors:
```python
f = pmd.openmm.load_topology(init.topology, system=system)
f.save('PEG2_parmed.top', overwrite=True)
```
Yields the following in [ default ] before:
```
[ defaults ]
; nbfunc        comb-rule       gen-pairs       fudgeLJ fudgeQQ
1               2               yes             0.5          0.8333
```
and after:
```
[ defaults ]
; nbfunc        comb-rule       gen-pairs       fudgeLJ fudgeQQ
1               2               no              1            0.8333
```

Looking forward to hearing from you and I apologize for posting many questions recently.

**Comment by mattwthompson** [comment]
Your exceptions, or at least some of them, aren't being built. Somebody else will have to explain why but that's probably the main cause of your energy differences.

What you see as a difference in `fudgeLJ` is because ParmEd stores the 1-4 interactions as special pairs in the `[ pairs ]` directive instead of letting GROMACS compte them on the fly `gen-pairs = yes`. Notice that your `PEG2.top` tells GROMACS _which_ atoms interact with 1-4 scaling factors, but does not compute them. Which is fine. It's just that ParmEd _does_ fill in the parameters for AMBER-related reasons. 

Your original GROMACS topology file:
```
[ pairs ]
;    ai     aj funct         c0         c1         c2         c3
      2      6     1
      3      6     1
      4      6     1
...
```

vs ParmEd's
```
[ pairs ]
;    ai     aj funct         c0         c1         c2         c3
      2      6     1 0.298451000 0.076693987
      3      6     1 0.298451000 0.076693987
      4      6     1 0.298451000 0.076693987
      1      7     1 0.301123500 0.109305749
```

IIRC, roundtripping GROMACS files with ParmEd introduces this difference and does not necessarily indicate that something is wrong there.

A much more minor difference that I noticed is that the 1-4 scaling factor is rounded from 0.8333333333333334 (last digit is strange) to 0.8333. This is probably not [rounding](https://github.com/ParmEd/ParmEd/blob/252c5ba52ebcd57509feb525d6da9e00f9d1ffe3/parmed/gromacs/gromacstop.py#L1431) in ParmEd, at least at the writing step.

**Comment by SHervoe-Hansen** [comment]
I can try to double-check the topology in terms of bonding. However, I am confident it should be correct as the bonds defined in the force field file and the pdb file should throw an exception if not the same.

> 1-4 scaling factor is rounded from 0.8333333333333334 (last digit is strange)

That should be correct as it is also the scaling factor found in the [ff14SB force field for openMM](https://github.com/openmm/openmmforcefields/blob/7e2c769003195089938a134c44b3c81b05aeb624/amber/ffxml/protein.ff14SB.xml#L3445).

**Comment by mattwthompson** [comment]
The 12th or so digit probably doesn't matter, I shouldn't have commented about that. Rounding from that value to 0.3333 might matter, though.

**Comment by SHervoe-Hansen** [comment]
@peastman would you happen to have an idea as to what might be happening?

I promise I will get off your back for a little while then. :sweat_smile:

**Comment by peastman** [comment]
The two versions certainly create different systems.  To see this, I used XmlSerializer to write out the first force, which happens to be a HarmonicBondForce.

```python
print(mm.XmlSerializer.serialize(system.getForce(0)))
```

Here's the version created from the ForceField:

```xml
<?xml version="1.0" ?>
<Force forceGroup="0" name="HarmonicBondForce" type="HarmonicBondForce" usesPeriodic="0" version="2">
	<Bonds>
		<Bond d=".1415" k="267776" p1="0" p2="1"/>
		<Bond d=".1415" k="267776" p1="1" p2="2"/>
		<Bond d=".153" k="259408" p1="2" p2="8"/>
		<Bond d=".1415" k="267776" p1="8" p2="9"/>
		<Bond d=".1415" k="267776" p1="9" p2="10"/>
	</Bonds>
</Force>
```

And here's the version created from the top file:

```xml
<?xml version="1.0" ?>
<Force forceGroup="0" name="HarmonicBondForce" type="HarmonicBondForce" usesPeriodic="0" version="2">
	<Bonds>
		<Bond d=".1111" k="284512" p1="0" p2="1"/>
		<Bond d=".1111" k="284512" p1="0" p2="2"/>
		<Bond d=".1111" k="284512" p1="0" p2="3"/>
		<Bond d=".1415" k="267776" p1="0" p2="4"/>
		<Bond d=".1415" k="267776" p1="4" p2="5"/>
		<Bond d=".1111" k="284512" p1="5" p2="6"/>
		<Bond d=".1111" k="284512" p1="5" p2="7"/>
		<Bond d=".15300000000000002" k="259408.00000000003" p1="5" p2="8"/>
		<Bond d=".1111" k="284512" p1="8" p2="9"/>
		<Bond d=".1111" k="284512" p1="8" p2="10"/>
		<Bond d=".1415" k="267776" p1="8" p2="11"/>
		<Bond d=".1415" k="267776" p1="11" p2="12"/>
		<Bond d=".1111" k="284512" p1="12" p2="13"/>
		<Bond d=".1111" k="284512" p1="12" p2="14"/>
		<Bond d=".1111" k="284512" p1="12" p2="15"/>
	</Bonds>
</Force>
```

I believe the problem comes from the fact that the PDB and top files have the atoms in different orders.  Compare the first few atoms in each one:

```
ATOM      1  C1  PGH     1      47.090   8.000   7.260  0.00  0.00           C  
ATOM      2  O2  PGH     1      47.400   9.350   6.960  0.00  0.00           O  
ATOM      3  C3  PGH     1      47.130  10.220   8.040  0.00  0.00           C  
ATOM      4  H11 PGH     1      46.090   7.890   7.710  0.00  0.00           H  
```

Compared to

```
    1         CF      1    PGH     C1      1 -0.10000000  12.011000   ; qtot -0.100000
    2         HF      1    PGH    H11      2 0.09000000   1.008000   ; qtot -0.010000
    3         HF      1    PGH    H12      3 0.09000000   1.008000   ; qtot 0.080000
    4         HF      1    PGH    H13      4 0.09000000   1.008000   ; qtot 0.170000
```

ParmEd expects them to be in the same order.  It's mixing information from the two files, which leads to incorrect results.  It might be useful to add a sanity check for this.  It could detect that atom 2 is an oxygen in the PDB file but a hydrogen in the top file and throw an exception.

cc @swails

**Comment by SHervoe-Hansen** [comment]
That would certainly explain it!

>ParmEd expects them to be in the same order. 

This is only true for gromacs files correct? Also am I wrong to assume openMM's (and Amber's) approach is the matching of atom names in the PDB to the atom names in the force field for the various residues?

**Comment by peastman** [comment]
In all cases where you load multiple files, they should have the atoms in the same order.  For example, Amber prmtop and inpcrd files.  Or CHARMM psf and crd files.  If they order the atoms differently, you'll get wrong results.

**Comment by SHervoe-Hansen** [comment]
Sorry I think I explained my self a bit poorly: The ordering of atoms within a single residue does not matter if you use .xml and pdb files for openMM correct? So long all the atom names are there?

**Comment by SHervoe-Hansen** [comment]
I think I understand entirely what is happening now. I apologize for the confusion.

Also thank you for introducing me to `XmlSerializer.serialize()`. I managed to debug the rest of the force field using that.

In that regard I did notice that when using `XmlSerializer.serialize()` on the HarmonicBondForce that all bonds involving hydrogen were gone (as expected when using `constraints=app.HBonds`), however, this was not the case for the ParmEd created system. The number of harmonic bonds was independent of `constraints=app.HBonds` or `constraints=None`. I guess this should, however, go to the ParmEd GitHub if it is an issue. In contrast `system.getNumConstraints()` returned the same number of constraints for both systems.
